File: src/data/o1/Longest_Common_Suffix_Queries/Longest_Common_Suffix_Queries.java
import java.util.*;

class Solution {
    class TrieNode {
        TrieNode[] children;
        int bestIndex; // Index of the best word for this suffix

        TrieNode() {
            children = new TrieNode[26];
            bestIndex = -1;
        }
    }

    public int[] stringIndices(String[] wordsContainer, String[] wordsQuery) {
        TrieNode root = new TrieNode();

        // Build the Trie with reversed words from wordsContainer
        for (int i = 0; i < wordsContainer.length; i++) {
            String word = wordsContainer[i];
            String reversed = new StringBuilder(word).reverse().toString();
            TrieNode current = root;
            for (char c : reversed.toCharArray()) {
                int idx = c - 'a';
                if (current.children[idx] == null) {
                    current.children[idx] = new TrieNode();
                }
                current = current.children[idx];
                // Update bestIndex if necessary
                if (current.bestIndex == -1) {
                    current.bestIndex = i;
                } else {
                    String existingWord = wordsContainer[current.bestIndex];
                    String newWord = wordsContainer[i];
                    if (newWord.length() < existingWord.length() ||
                       (newWord.length() == existingWord.length() && i < current.bestIndex)) {
                        current.bestIndex = i;
                    }
                }
            }
        }

        int[] ans = new int[wordsQuery.length];
        Arrays.fill(ans, -1);

        // Process each query
        for (int i = 0; i < wordsQuery.length; i++) {
            String query = wordsQuery[i];
            String reversedQuery = new StringBuilder(query).reverse().toString();
            TrieNode current = root;
            int bestCandidate = -1;
            for (char c : reversedQuery.toCharArray()) {
                int idx = c - 'a';
                if (current.children[idx] == null) {
                    break;
                }
                current = current.children[idx];
                if (current.bestIndex != -1) {
                    bestCandidate = current.bestIndex;
                }
            }
            ans[i] = bestCandidate;
        }

        return ans;
    }
}


public class Longest_Common_Suffix_Queries {
public static void runTests(Solution solution) {
    // Test case 1: Simple case with exact matches on suffix
    String[] wordsContainer1 = {"abcd", "bcd", "xbcd"};
    String[] wordsQuery1 = {"cd", "bcd", "xyz"};
    int[] expected_result1 = {1, 1, 1};
    assert Arrays.equals(solution.stringIndices(wordsContainer1, wordsQuery1), expected_result1) : "Test case 1 failed. Expected " + Arrays.toString(expected_result1);

    // Test case 2: Multiple words sharing the same longest suffix
    String[] wordsContainer2 = {"hello", "fellow", "mellow", "yellow"};
    String[] wordsQuery2 = {"below"};
    int[] expected_result2 = {1};  // "fellow" has the same suffix "ellow" and is smallest in length
    assert Arrays.equals(solution.stringIndices(wordsContainer2, wordsQuery2), expected_result2) : "Test case 2 failed. Expected " + Arrays.toString(expected_result2);

    // Test case 3: No common suffix found
    String[] wordsContainer3 = {"cat", "dog", "bird"};
    String[] wordsQuery3 = {"fish", "whale"};
    int[] expected_result3 = {0, 0};  // No matching suffixes
    assert Arrays.equals(solution.stringIndices(wordsContainer3, wordsQuery3), expected_result3) : "Test case 3 failed. Expected " + Arrays.toString(expected_result3);

    // Test case 4: Suffix matches from the beginning of wordsContainer
    String[] wordsContainer4 = {"apple", "ample", "simple", "temple"};
    String[] wordsQuery4 = {"ample"};
    int[] expected_result4 = {1};  // "ample" is an exact match in wordsContainer
    assert Arrays.equals(solution.stringIndices(wordsContainer4, wordsQuery4), expected_result4) : "Test case 4 failed. Expected " + Arrays.toString(expected_result4);

    // Test case 5: Same length suffixes, but choose earlier in wordsContainer
    String[] wordsContainer5 = {"catch", "match", "latch"};
    String[] wordsQuery5 = {"batch"};
    int[] expected_result5 = {0};  // "catch" has the same suffix "atch" as the other words but occurs earlier
    assert Arrays.equals(solution.stringIndices(wordsContainer5, wordsQuery5), expected_result5) : "Test case 5 failed. Expected " + Arrays.toString(expected_result5);

    // Test case 6: Mixed case with varying suffix lengths
    String[] wordsContainer6 = {"abcdefgh", "poiuygh", "ghghgh"};
    String[] wordsQuery6 = {"gh", "acbfgh", "acbfegh"};
    int[] expected_result6 = {2, 0, 2};  // Matches "ghghgh" for "gh", and "poiuygh" for "acbfegh"
    assert Arrays.equals(solution.stringIndices(wordsContainer6, wordsQuery6), expected_result6) : "Test case 6 failed. Expected " + Arrays.toString(expected_result6);

    // Test case 7: Single character suffix matches
    String[] wordsContainer7 = {"a", "b", "c", "d"};
    String[] wordsQuery7 = {"e", "a", "z"};
    int[] expected_result7 = {0, 0, 0};  // Only "a" matches, the rest have no match
    assert Arrays.equals(solution.stringIndices(wordsContainer7, wordsQuery7), expected_result7) : "Test case 7 failed. Expected " + Arrays.toString(expected_result7);
}


    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 5000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/o1/Longest_Common_Suffix_Queries/Longest_Common_Suffix_Queries.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> stringIndices(vector<string>& wordsContainer, vector<string>& wordsQuery) {
        // Build the Trie with reversed words from wordsContainer
        TrieNode* root = new TrieNode();
        int n = wordsContainer.size();
        vector<int> lengths(n, 0);
        for(int i = 0; i < n; ++i){
            lengths[i] = wordsContainer[i].size();
            string reversed_word = wordsContainer[i];
            reverse(reversed_word.begin(), reversed_word.end());
            insert(root, reversed_word, i, lengths);
        }
        
        // Process each query
        vector<int> ans;
        for(auto &query : wordsQuery){
            string reversed_query = query;
            reverse(reversed_query.begin(), reversed_query.end());
            int best_index = -1;
            TrieNode* current = root;
            for(char c : reversed_query){
                if(!current->children[c - 'a']) break;
                current = current->children[c - 'a'];
                if(current->best_index != -1){
                    best_index = current->best_index;
                }
            }
            ans.push_back(best_index);
        }
        
        // Clean up Trie memory
        deleteTrie(root);
        
        return ans;
    }

private:
    struct TrieNode {
        TrieNode* children[26];
        int best_index;
        TrieNode() : best_index(-1) {
            for(int i = 0; i < 26; ++i) children[i] = nullptr;
        }
    };
    
    // Function to determine if index1 is better than index2
    bool isBetter(int index1, int index2, const vector<int>& lengths){
        if(lengths[index1] < lengths[index2]) return true;
        if(lengths[index1] == lengths[index2] && index1 < index2) return true;
        return false;
    }
    
    // Insert a reversed word into the Trie
    void insert(TrieNode* root, const string& word, int index, const vector<int>& lengths){
        TrieNode* node = root;
        for(char c : word){
            if(!node->children[c - 'a']){
                node->children[c - 'a'] = new TrieNode();
            }
            node = node->children[c - 'a'];
            if(node->best_index == -1 || isBetter(index, node->best_index, lengths)){
                node->best_index = index;
            }
        }
    }
    
    // Delete the Trie to free memory
    void deleteTrie(TrieNode* node){
        if(!node) return;
        for(int i = 0; i < 26; ++i){
            if(node->children[i]){
                deleteTrie(node->children[i]);
            }
        }
        delete node;
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Simple case with exact matches on suffix
    std::vector<std::string> wordsContainer = {"abcd", "bcd", "xbcd"};
    std::vector<std::string> wordsQuery = {"cd", "bcd", "xyz"};
    std::vector<int> expected_result = {1, 1, 1};
    assert(solution.stringIndices(wordsContainer, wordsQuery) == expected_result && "Test case 1 failed.");

    // Test case 2: Multiple words sharing the same longest suffix
    wordsContainer = {"hello", "fellow", "mellow", "yellow"};
    wordsQuery = {"below"};
    expected_result = {1};  // "fellow" has the same suffix "ellow" and is smallest in length
    assert(solution.stringIndices(wordsContainer, wordsQuery) == expected_result && "Test case 2 failed.");

    // Test case 3: No common suffix found
    wordsContainer = {"cat", "dog", "bird"};
    wordsQuery = {"fish", "whale"};
    expected_result = {0, 0};  // No matching suffixes
    assert(solution.stringIndices(wordsContainer, wordsQuery) == expected_result && "Test case 3 failed.");

    // Test case 4: Suffix matches from the beginning of wordsContainer
    wordsContainer = {"apple", "ample", "simple", "temple"};
    wordsQuery = {"ample"};
    expected_result = {1};  // "ample" is an exact match in wordsContainer
    assert(solution.stringIndices(wordsContainer, wordsQuery) == expected_result && "Test case 4 failed.");

    // Test case 5: Same length suffixes, but choose earlier in wordsContainer
    wordsContainer = {"catch", "match", "latch"};
    wordsQuery = {"batch"};
    expected_result = {0};  // "catch" has the same suffix "atch" as the other words but occurs earlier
    assert(solution.stringIndices(wordsContainer, wordsQuery) == expected_result && "Test case 5 failed.");

    // Test case 6: Mixed case with varying suffix lengths
    wordsContainer = {"abcdefgh", "poiuygh", "ghghgh"};
    wordsQuery = {"gh", "acbfgh", "acbfegh"};
    expected_result = {2, 0, 2};  // "flow" for "low", and "snow" for "snowfall"
    assert(solution.stringIndices(wordsContainer, wordsQuery) == expected_result && "Test case 6 failed.");

    // Test case 7: Single character suffix matches
    wordsContainer = {"a", "b", "c", "d"};
    wordsQuery = {"e", "a", "z"};
    expected_result = {0, 0, 0};  // Only "a" matches, the rest have no match
    assert(solution.stringIndices(wordsContainer, wordsQuery) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 327680; ++i) {
        testSolution(solution);  
    }
    return 0;
}
